src/mumble/PulseAudio.h

```diff
@@ -80,6 +93,67 @@ class PulseAudioSystem : public QObject {
+		void (*pa_context_disconnect)(pa_context *c);
+		pa_operation *(*pa_context_subscribe)(pa_context *c, pa_subscription_mask_t m, pa_context_success_cb_t cb, void *userdata);
+		pa_context_state_t (*pa_context_get_state)(const pa_context *c);
+		pa_operation *(*pa_context_get_server_info)(pa_context *c, pa_server_info_cb_t cb, void *userdata);
+		pa_operation *(*pa_context_get_sink_info_by_name)(pa_context *c, const char *name, pa_sink_info_cb_t cb, void *userdata);
+		pa_operation *(*pa_context_get_sink_info_list)(pa_context *c, pa_sink_info_cb_t cb, void *userdata);
+		pa_operation *(*pa_context_get_sink_input_info_list)(pa_context *c, pa_sink_input_info_cb_t cb, void *userdata);
+		pa_operation *(*pa_context_get_source_info_list)(pa_context *c, pa_source_info_cb_t cb, void *userdata);
+		pa_operation *(*pa_context_set_sink_input_volume)(pa_context *c, uint32_t idx, const pa_cvolume *volume, pa_context_success_cb_t cb, void *userdata);
+		void (*pa_context_set_state_callback)(pa_context *c, pa_context_notify_cb_t cb, void *userdata);
+		void (*pa_context_set_subscribe_callback)(pa_context *c, pa_context_subscribe_cb_t cb, void *userdata);
+
+		pa_stream *(*pa_stream_new)(pa_context *c, const char *name, const pa_sample_spec *ss, const pa_channel_map *map);
+		void (*pa_stream_unref)(pa_stream *s);
+		int (*pa_stream_connect_playback)(pa_stream *s, const char *dev, const pa_buffer_attr *attr, pa_stream_flags_t flags, const pa_cvolume *volume, pa_stream *sync_stream);
+		int (*pa_stream_connect_record)(pa_stream *s, const char *dev, const pa_buffer_attr *attr, pa_stream_flags_t flags);
+		int (*pa_stream_disconnect)(pa_stream *s);
+		int (*pa_stream_peek)(pa_stream *p, const void **data, size_t *nbytes);
+		int (*pa_stream_write)(pa_stream *p, const void *data, size_t nbytes, pa_free_cb_t free_cb, int64_t offset, pa_seek_mode_t seek);
+		int (*pa_stream_drop)(pa_stream *p);
+		pa_operation *(*pa_stream_cork)(pa_stream *s, int b, pa_stream_success_cb_t cb, void *userdata);
+		pa_stream_state_t (*pa_stream_get_state)(const pa_stream *p);
+		pa_context *(*pa_stream_get_context)(const pa_stream *p);
+		const pa_sample_spec *(*pa_stream_get_sample_spec)(pa_stream *s);
+		const pa_channel_map *(*pa_stream_get_channel_map)(pa_stream *s);
+		const pa_buffer_attr *(*pa_stream_get_buffer_attr)(pa_stream *s);
+		void (*pa_stream_set_state_callback)(pa_stream *s, pa_stream_notify_cb_t cb, void *userdata);
+		void (*pa_stream_set_read_callback)(pa_stream *p, pa_stream_request_cb_t cb, void *userdata);
+		void (*pa_stream_set_write_callback)(pa_stream *p, pa_stream_request_cb_t cb, void *userdata);
+
+		pa_operation *(*pa_ext_stream_restore_read)(pa_context *c, pa_ext_stream_restore_read_cb_t cb, void *userdata);
+		pa_operation *(*pa_ext_stream_restore_write)(pa_context *c, pa_update_mode_t mode, const pa_ext_stream_restore_info data[], unsigned n, int apply_immediately, pa_context_success_cb_t cb, void *userdata);
```

Comment:
Imo we should create a dedicated struct for this and make that a member variable of the `PulseAudioSystem` class. Seems like a cleaner solution to me than having a bunch of "free" function pointers as class members.

And while we're one it, we could turn that into a proper calss that also automatically takes care of loading the respective functions via the library. That'd then make sure that we have a separation between the library wrapper and the actual code that wants to use it.

Reply:
The problem is that we would have to specify the class object every time we want to call a PulseAudio function.

Reply:
so what? I don't see this as a problem. In fact we can even drop the `pa_` prefix for the functions in that case so the overall overhead in terms of writing should be rather minimal. Plus this change can probably be done by search&replace anyways

Reply:
Yep, excellent idea.